Revert "Remove TestSearchCmdOptions test"

thaJeztah opened the pull request:
This reverts commit 21e662c77468d1e3fc4efaa049021e2ad0e49b4d.

see https://github.com/moby/moby/pull/39257#issuecomment-494789534 😓 

Also made some additional improvements;

- removed checks for client-only functionality (as we're testing against a fixed version of the cli)
- improved some boundary checks and small refactor
- enabled the tests on windows

@cpuguy83 @kolyshkin PTAL (I'll close if we're ok with the PR being merged 🤗)

thaJeztah committed:
Revert "Remove TestSearchCmdOptions test"

This reverts commit 21e662c77468d1e3fc4efaa049021e2ad0e49b4d.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

cpuguy83 commented:
🤷‍♂ 

Maybe it'd be nice to be able to stub out hub in the tests so we can control the result values.
I think we should keep the test around or we should deprecate the functionality, at least in the engine since it should really be client side anyway.

thaJeztah commented:
> Maybe it'd be nice to be able to stub out hub in the tests so we can control the result values.

Yes, I actually  went looking for that (also to check if the generated URLs/Queries look ok); there's some unit tests in https://github.com/moby/moby/blob/master/daemon/images/image_search_test.go

tonistiigi commented:
The CI for this seemed to fail on the stars count vs unfiltered check that should be quite consistent.

```
assertion failed: expression is false: strings.Count(outSearchCmdStars, "[OK]") <= strings.Count(outSearchCmd, "[OK]"): The quantity of images with stars should be less than that of all images: NAME                        DESCRIPTION                                     STARS     OFFICIAL   AUTOMATED
11:50:46 busybox                     Busybox base image.                             1584      [OK]       
11:50:46 progrium/busybox                                                            69                   [OK]
11:50:46 hypriot/rpi-busybox-httpd   Raspberry Pi compatible Docker Image with ...   46                   
11:50:46 radial/busyboxplus          Full-chain, Internet enabled, busybox made...   24                   [OK]
11:50:46 hypriot/armhf-busybox       Busybox base image for ARM.                     11                   
11:50:46 arm32v7/busybox             Busybox base image.                             7                    
11:50:46 yauritux/busybox-curl       Busybox with CURL                               5                    
11:50:46 armhf/busybox               Busybox base image.                             4                    
11:50:46 arm64v8/busybox             Busybox base image.                             3                    
11:50:46 prom/busybox                Prometheus Busybox Docker base images           2                    [OK]
11:50:46 aarch64/busybox             Busybox base image.                             2                    
11:50:46 s390x/busybox               Busybox base image.                             2                    
11:50:46 odise/busybox-curl                                                          2                    [OK]
11:50:46 i386/busybox                Busybox base image.                             2                    
11:50:46 p7ppc64/busybox             Busybox base image for ppc64.                   2                    
11:50:46 armel/busybox               Busybox base image.                             2                    
11:50:46 sequenceiq/busybox                                                          2                    [OK]
11:50:46 arm32v6/busybox             Busybox base image.                             2                    
11:50:46 
11:50:46 
11:50:46 ----------------------------------------------------------------------
```

So if we put it back I'd like to log out the actual comparison values to see what was wrong with it. Atm only one side is printed.

kolyshkin commented:
I'm against putting this back -- the functionality is pretty simple, there's definitely no need to test it using docker hub.

tonistiigi commented:
search does not have a reference implementation and is directly a hub feature afaik. It's important to make sure we don't ship a version that has a broken search. This is as important to changes coming to engine as it is for hub changes. Can't have a situation where both are running tests only against their mocks and think both are doing the correct thing.

kolyshkin commented:
> This is as important to changes coming to engine as it is for hub changes. 

Moby CI is probably not the best place to test Docker Hub; from the other side there's nothing wrong with it, assuming the Hub won't break.

kolyshkin reviewed (approved):
LGTM

thaJeztah committed:
TestSearchCmdOptions: remove cli-only checks

Both `--help` and `--no-trunc` are implemented in the CLI. There's
no need to verify them here because the integration tests use a
fixed version of the CLI.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestSearchCmdOptions: remove checks for deprecated flags

The `--stars` and `--automated` flags have been deprecated, and were
replaced by `--filter stars=xx` and `--filter is-automated=true`.

Integration tests run with a fixed version of the CLI, and the new
(`--filter`) option is already tested in this test, so there's no need
to verify the old flags.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestSearchStarsOptionWithWrongParameter: remove checks for deprecated flags

The `--stars` flag was deprecated, and was replaced by `--filter stars=xx`

Integration tests run with a fixed version of the CLI, and the new
(`--filter`) option is already tested in this test, so there's no need
to verify the old flags.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
TestSearchWithLimit: slight refactor and improve boundary checks

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
integration-cli: also run Docker Hub search tests on Windows

The API does not filter images on platform, so searching on
Windows should work as well.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Added some commits with additional improvements;

- removed checks for client-only functionality (as we're testing against a fixed version of the cli)
- improved some boundary checks and small refactor
- enabled the tests on windows

thaJeztah commented:
Janky failure should be ignored;

```
11:21:38 The following new tests were added to integration-cli:
11:21:38 
11:21:38 +func (s *DockerSuite) TestSearchCmdOptions(c *check.C) {
11:21:38 
11:21:38 integration-cli is deprecated. Please add an API integration test to
11:21:38 ./integration/COMPONENT/. See ./TESTING.md for more details.
```

thaJeztah commented:
Tests completed successfully on other jobs;

```
12:06:11 PASS: docker_cli_search_test.go:35: DockerSuite.TestSearchCmdOptions	2.192s
12:06:12 PASS: docker_cli_search_test.go:12: DockerSuite.TestSearchOnCentralRegistry	0.350s
12:06:12 PASS: docker_cli_search_test.go:63: DockerSuite.TestSearchOnCentralRegistryWithDash	0.380s
12:06:12 PASS: docker_cli_search_test.go:17: DockerSuite.TestSearchStarsOptionWithWrongParameter	0.060s
12:06:13 PASS: docker_cli_search_test.go:68: DockerSuite.TestSearchWithLimit	1.347s
```

tonistiigi reviewed (approved):
If we have some good monitoring on master we could run tests like this only in master and not in PRs to reduce the risk of outside errors marking PRs red.

cpuguy83 commented:
Honestly I'd be up for receiving a daily digest of failed runs on master.

thaJeztah commented:
I can redirect my email notifications to you 🤗

cpuguy83 commented:
"mark as spam"

kolyshkin commented:
LGTM

kolyshkin commented:
LGTM
